inc/appliance.class.php

```diff
@@ -0,0 +1,602 @@
+    *
+    * @param boolean $all All types  or only allowed ones (defauts to false)
+    *
+    * @return array of types
+   **/
+   static function getTypes($all = false) {
+      global $CFG_GLPI;
+
+      $types = $CFG_GLPI['appliance_types'];
+
+      if ($all) {
+         return $types;
+      }
+
+      // Only allowed types
+      foreach ($types as $key => $type) {
+         if (!($item = getItemForItemtype($type))) {
+            continue;
+         }
+
+         if (!$item->canView()) {
+            unset($types[$key]);
+         }
+      }
+      return $types;
+   }
+
+
+   function getSpecificMassiveActions($checkitem = null) {
+
+      $isadmin = static::canUpdate();
+      $actions = parent::getSpecificMassiveActions($checkitem);
+
+      if ($_SESSION['glpiactiveprofile']['interface'] == 'central') {
+         if ($isadmin) {
+            $actions['Appliance'.MassiveAction::CLASS_ACTION_SEPARATOR.'install'] = _x('button', 'Associate');
+            $actions['Appliance'.MassiveAction::CLASS_ACTION_SEPARATOR.'uninstall'] = _x('button', 'Dissociate');
+         }
+      }
+      return $actions;
+   }
+
+
+   static function showMassiveActionsSubForm(MassiveAction $ma) {
+
+      switch ($ma->getAction()) {
+         case 'add_item':
+            self::dropdownMA([]);
+            echo "&nbsp;". Html::submit(_x('button', 'Post'), ['name' => 'massiveaction']);
+            return true;
+
+         case "install" :
+            Dropdown::showSelectItemFromItemtypes([
+               'items_id_name' => 'item_item',
+               'itemtype_name' => 'typeitem',
+               'itemtypes'     => self::getTypes(true),
+               'checkright'    => true
+            ]);
+            echo Html::submit(_x('button', 'Post'), ['name' => 'massiveaction']);
+            return true;
+
+         case "uninstall" :
+            Dropdown::showSelectItemFromItemtypes([
+               'items_id_name' => 'item_item',
+               'itemtype_name' => 'typeitem',
+               'itemtypes'     => self::getTypes(true),
+               'checkright'    => true
+            ]);
+            echo Html::submit(_x('button', 'Post'), ['name' => 'massiveaction']);
+            return true;
+      }
+      return parent::showMassiveActionsSubForm($ma);
+   }
+
+
+   static function processMassiveActionsForOneItemtype(MassiveAction $ma, CommonDBTM $item, array $ids) {
+
+      $appliance_item = new Appliance_Item();
+
+      switch ($ma->getAction()) {
+         case "add_item":
+            $input = $ma->getInput();
+            foreach ($ids as $id) {
+               $input = [
+                  'appliances_id'   => $input['appliances_id'],
+                  'items_id'        => $id,
+                  'itemtype'        => $item->getType()
+               ];
+               if ($appliance_item->can(-1, CREATE, $input) && !$appliance_item->getFromDBByCrit($input)) {
+                  if ($appliance_item->add($input)) {
```

Comment:
If already exists, it could be considered as a success.
```suggestion
               if ($appliance_item->can(-1, CREATE, $input)) {
                  if ($appliance_item->getFromDBByCrit($input) || $appliance_item->add($input)) {
```